File: osf/libs/entities/src/utc-iso-timestamp.ts
/**
 * Canonical ISO-8601 UTC timestamps with millisecond precision and `Z` suffix.
 * Same string shape as `Date.prototype.toISOString()` and Arduino `OSF_MultiSensor_R4WiFi` v1.1.6+ payload `timestamp`.
 */
export function utcIsoTimestampMs(date: Date = new Date()): string {
  return date.toISOString();
}

File: session_manager/utils/utc_iso_timestamp.py
"""Canonical UTC ISO-8601 with millisecond precision and Z suffix (aligned with OSF `utcIsoTimestampMs` / Arduino v1.1.6+)."""

from datetime import datetime, timezone


def utc_iso_timestamp_ms() -> str:
    """Return e.g. ``2026-03-31T12:00:00.123Z``."""
    dt = datetime.now(timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S") + f".{dt.microsecond // 1000:03d}Z"
